### ISSUE TITLE
UNION/UNION-ALL priority too high compared to WITH

### ISSUE THREAD

**Issue by emidln** [jkk/honeysql#165]
https://github.com/jkk/honeysql/commit/6a0eacdd5fc301055b0e18be31a034c7304ea0a1 breaks a CTE used to feed data into queries that are UNION'd. This test case is a little nonsensical given that the UNIONs don't actually use the data provided by the CTE, but this is a simplified example of a query I have that now breaks having upgraded from 0.8.1 to 0.8.2.

```
diff --git a/test/honeysql/format_test.cljc b/test/honeysql/format_test.cljc
index 3e6faf7..f501fd0 100644
--- a/test/honeysql/format_test.cljc
+++ b/test/honeysql/format_test.cljc
@@ -121,3 +121,10 @@
            (format {:select [:foo]
                     :from [:bar]
                     :where [:= [:mod :col1 4] [:+ :col2 4]]})))))
+
+(deftest union-with-cte
+  (is (= (format {:union [{:select [:foo] :from [:bar1]}
+                          {:select [:foo] :from [:bar2]}]
+                  :with [[[:bar {:columns [:spam :eggs]}]
+                          {:values [[1 2] [3 4] [5 6]]}]]})
+         ["WITH bar (spam, eggs) AS (VALUES (?, ?), (?, ?), (?, ?)) SELECT foo FROM bar1 UNION SELECT foo FROM bar2" 1 2 3 4 5 6])))
```

I think if we prioritized WITH such that it appears before UNION/UNION-ALL but still before ORDER BY, this would work again.

**Comment by MichaelBlume** [comment]
I'm feeling a bit out of my depth. @seancorfield, care to weigh in?

**Comment by seancorfield** [comment]
Yeah, I don't use CTE's myself but I follow @emidln's reasoning and since the associated PR #166 only affects the interaction of `with`/`union` I think it's the right thing to do.

**Comment by seancorfield** [comment]
To expand on that a bit... I originally introduced the specific priorities for `union` in 0.8.1 via a PR since they mostly did the wrong thing. Then in 0.8.2 there was another PR to correct a bug I hadn't noticed (but, ironically, ran into later!) with parentheses and `union`. Those two changes weren't tested widely enough so the fix in 0.8.2 broke yet another case (above) that really only worked by accident in 0.8.1.

**Comment by MichaelBlume** [comment]
Thanks, I'll merge the PR =)

**Comment by MichaelBlume** [comment]
This is now fixed in 0.9.0. Thanks =)
